include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td

```diff
@@ -282,6 +282,16 @@ def AndOfPad : Pat <
   (MoveNameHint $old, (PadPrimOp (AndPrimOp (TailPrimOp $x, (TypeWidthAdjust32 $x, $y)), $y), $n)),
   [(KnownWidth $x), (UIntType $x), (EqualTypes $x, $pad)]>;
 
+/// cvt is adding a 0, which will pass through the and
+/// and(cvt(x:U), const) -> cat(0, and(x,trunc(const))
+def AndCvtU : Pat <
+  (AndPrimOp:$old (CvtPrimOp $x), (ConstantOp:$ocst $cst)),
+  (MoveNameHint $old, (CatPrimOp 
+    (NativeCodeCall<"$_builder.create<ConstantOp>($0.getLoc(), APSInt(APInt()))"> $ocst),
+    (AndPrimOp $x, (NativeCodeCall<"$_builder.create<ConstantOp>($0.getLoc(), cast<IntType>($0.getType()), $1.getValue().trunc($1.getValue().getBitWidth()-1).zext(*cast<IntType>($0.getType()).getWidth()))"> $x, $cst))
+  )),
+  [(KnownWidth $x), (UIntType $x)]>;
```

Comment:
There is a variant of this pattern that applies for all of `OrCvtU`, `XorCvtU`, and `NotCvtU`. (Anytime there is a `cvt` whose user is an op which does an `SInt` -> `UInt` conversion.) The patterns should be the following:

- `or(cvt(x:U), const) -> cat(head(const, 1), or(x, trunc(const))`
- `xor(cvt(x:U), const) -> cat(head(const, 1), xor(x, trunc(const)))`
- `not(cvt(x:U)) -> cat(1, not(x))`

Did you experiment with these, too?

Reply:
I didn't see them in the code I was looking at.  There were a lot of add/sum after a cvt which might be able to be narrowed.

Then would be a good thing to add at some point.